### PR TITLE
Add grouping by territory to LanguagePickerModal

### PR DESCRIPTION
**Pull request by jsnajdr**
Groups the languages by territory where the language is spoken (Africa, America, Europe, ...) and shows tabs in the `LanguagePickerModal`:

language-territories

Very similar to the Facebook language selection dialog.

@folletto The original design doesn't have the 'Popular languages' and 'All languages' tabs, but they seem to be pretty useful nevertheless. What do you think? A little trouble with so many tabs is that the dialog needs to be at least 1000px wide to accomodate them all. Otherwise, they collapse into a dropdown menu (a standard behavior of `SectionNavTabs`):

language-territories-collapsed

### CODE REVIEW

**Comment by folletto** [comment]
> The original design doesn't have the 'Popular languages' and 'All languages' tabs, but they seem to be pretty useful nevertheless.

These weren't added on purpose because:

1. As you correctly mention, the list gets too long, and adds an extra click sooner
2. The suggested languages bar already covers the needs there
3. We have search

In short: let's not add either "All Languages" nor "Popular Languages".

However, your comment highlights me that could be useful if we pre-select the right tab using geolocation. In this way one would get their browser preferences in "Suggested Languages" and the geolocation options in the tab, covering all bases.

**Comment by jsnajdr** [comment]
> In short: let's not add either "All Languages" nor "Popular Languages".

There are three languages that are not currently associated with any territory and that we must handle somehow before removing "All Languages":
- Esperanto
- Interlingua
- Yiddish

Yiddish doesn't have any territory association in the CLDR data, but I guess it can be assigned either to Western or Eastern Europe? @yoavf might know more.

But the other two? Make them discoverable only with search?

Pre-selecting the tab by geolocation is a great idea :)

**Comment by folletto** [comment]
Ok, I haven't expected these. Nice find.

I'd think Esperanto and Interlingua are good to be just searchable, as probably whoever knows these searches knows they aren't regional languages.

I've not enough knowledge on Yiddish. Judging from its recognition state on Wikipedia, I could guess to localize in Europe, but there's probably a reason why it's in "All Languages" and I'd follow that. We can probably show it in all tabs, so it's always available. We might do the same for the other two languages if you think it's a viable option for these too.

**Comment by jsnajdr** [comment]
Yiddish is currently only in "All Languages" solely because I didn't put it in any other group yet. The CLDR database doesn't have territory information for it, but so it doesn't for many other European regional languages like Occitan, Napolitan or Venetian. I had to sort a few of these manually.

If there is no strong opposition, I'll sort Yiddish to "Eastern Europe" and make Esperanto and Interlingua only searchable.

**Comment by yoavf** [comment]
That's an interesting discussion. Yiddish originates in eastern Europe, but it's biggest communities of speakers now are in Israel and the United States.

I think geographically splitting the languages might not be the best solution. Is "Spain" going to be listed under Europe? Is that where Spanish speaking Americans should go look for it?

**Comment by yoavf** [comment]
At least I'd argue that we keep the "Popular" tab, and maybe call it "Suggested". We should prefill that with the user's browser language settings (all options listed), add then all major languages spoken in the territory (i.e: for [Israel](http://www.unicode.org/cldr/charts/latest/supplemental/territory_language_information.html), that could be Hebrew, English, Arabic, Russian), and possibly a few other popular choices. 
English should always be on that list.

**Comment by folletto** [comment]
> At least I'd argue that we keep the "Popular" tab, and maybe call it "Suggested". We should prefill that with the user's browser language settings (all options listed), 

We already have the "Suggested Languages" for this, which is more effective as always visible regardless of the tab selection:

screen shot 2017-05-03 at 13 44 44


> add all major languages spoken in the territory (i.e: for Israel, that could be Hebrew, English, Arabic, Russian), and possibly a few other popular choices.

How does this solve Esperanto, Interlingua and Yiddish? It's the same issue of territory we have. 

I agree it's not "correct" to split languages by territory, but people do that. That's an intuitive way to navigate languages. 

> If there is no strong opposition, I'll sort Yiddish to "Eastern Europe" and make Esperanto and Interlingua only searchable.

Given the feedback above, let's make these three only searchable then for now, and we iterate from there once this lands.

**Comment by jsnajdr** [comment]
> I think geographically splitting the languages might not be the best solution. Is "Spanish" going to be listed under Europe? Is that where Spanish speaking Americans should go look for it?

It's perfectly OK to list a language in multiple territories, if we feel it's the right thing to do. My PR already does that with Spanish, Arabic and Kurdish.

**Comment by folletto** [comment]
Yep good point. Plus... the "Spanish" in Europe should be `es_ES`, while "Spanish" in America should be `es_LA` or `es_MX`, no? Isn't that a problem already solved?

**Comment by jsnajdr** [comment]
> Yep good point. Plus... the "Spanish" in Europe should be es_ES, while "Spanish" in America should be es_LA or es_MX, no? Isn't that a problem already solved?

The language list in `config/_shared.json` includes only `es` and `es-cl` (Chile) locales. The best thing to do seems to be to include `es` in both "Americas" and "Western Europe". If there was a `es-la` locale, then it would be the choice for "Americas" and `es` would be only in "Western Europe".

**Comment by folletto** [comment]
It's ok. Let's get there by steps. Once we have this merged we can start working on allocation right? ;)

Maybe add to the docs how to shuffle these and duplicate entries?

**Comment by yoavf** [comment]
> the "Spanish" in Europe should be es_ES, while "Spanish" in America should be es_LA or es_MX, no? Isn't that a problem already solved?

Not right now. On calypso the only Spanish that makes sense right now is just `es` - all the other variants won't be translated. That's why it the "Popular Locales" list made sense - those were all 100% translated. A Chilean will probably be better served by choosing `es` instead of `es-cl`

**Comment by folletto** [comment]
Ok, thanks for the clarification! Let's duplicate them in all locales where it makes sense then. 

I'd suggest do do a taxonomy review for these in a separate PR maybe?

**Comment by jsnajdr** [comment]
> However, your comment highlights me that could be useful if we pre-select the right tab using geolocation.

Is there any good default tab in case the geolocation information is not available? Both the `public-api.wordpress.com/geo` REST endpoint and the browser Geolocation API can fail, and both are async -- it  takes a while after opening the modal before the result arrives.

**Comment by folletto** [comment]
As before, let's start with something simple and work from there. What about choosing the largest population area as default, i.e. Asia?

File: client/components/language-picker/modal.jsx [hunk 1]
```diff
@@ -2,19 +2,116 @@
...
+			'NI', 'PA', 'SV', 'AG', 'AI', 'AW', 'BB', 'BL', 'BQ', 'BS',
+			'CU', 'CW', 'DM', 'DO', 'GD', 'GP', 'HT', 'JM', 'KN', 'KY',
+			'LC', 'MF', 'MQ', 'MS', 'PR', 'SX', 'TC', 'TT', 'VC', 'VG',
+			'VI', 'AR', 'BO', 'BR', 'CL', 'CO', 'EC', 'FK', 'GF', 'GY',
+			'PE', 'PY', 'SR', 'UY', 'VE'
+		],
+	},
+	{
+		id: 'asia-pacific',
+		name: translate => translate( 'Asia-Pacific' ),
```
Comment (a8ci18n) [inline client/components/language-picker/modal.jsx]: Hi! I've found a possible matching string that has already been translated **87** times:
`translate( 'Pacific' )` *ES Score: 9*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/components/language-picker/modal.jsx [hunk 2]
```diff
@@ -2,19 +2,116 @@
...
+			'CV', 'GH', 'GM', 'GN', 'GW', 'LR', 'ML', 'MR', 'NE', 'NG',
+			'SH', 'SL', 'SN', 'TG', 'AO', 'CD', 'CF', 'CG', 'CM', 'GA',
+			'GQ', 'ST', 'TD', 'BI', 'DJ', 'ER', 'ET', 'KE', 'KM', 'MG',
+			'MU', 'MW', 'MZ', 'RE', 'RW', 'SC', 'SO', 'SS', 'TZ', 'UG',
+			'YT', 'ZM', 'ZW', 'BW', 'LS', 'NA', 'SZ', 'ZA'
+		],
+	},
+	{
+		id: 'americas',
+		name: translate => translate( 'Americas' ),
```
Comment (a8ci18n) [inline client/components/language-picker/modal.jsx]: Hi! I've found a possible matching string that has already been translated **89** times:
`translate( 'America' )` *ES Score: 21*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

**Comment by jsnajdr** [comment]
Pushed an update where:
- the "All Languages" and "Popular Languages" tabs are removed. Only the regional groupings remain
- the default tab is "Asia-Pacific" -- largest population area
- geolocation is used to select the default tab. As the geolocation API returns a country code, had to fetch the mapping to regions from the CLDR `territoryContainment` data.
- if geolocation is not available, the default tab is "Asia-Pacific" -- largest population area

**Comment by alisterscott** [comment]
Closing as this PR is stale - please reopen if necessary

**Pull request closed without merging**